### PR TITLE
buildGoModule: remove modSha256

### PR DESCRIPTION
**Pull request by zowoq**
This is dead code but I'll wait until we're closer to 20.09 before merging.

### CODE REVIEW

File: pkgs/development/go-modules/generic/old.nix [hunk 1]
```diff
@@ -1,229 +0,0 @@
...
-      export GOPATH="$TMPDIR/go"
-      mkdir -p "''${GOPATH}/pkg/mod/cache/download"
-      cd "${modRoot}"
-      runHook postConfigure
-    '';
-
-    buildPhase = args.modBuildPhase or ''
-      runHook preBuild
-
-      go mod download
```
Comment (marsam) [inline pkgs/development/go-modules/generic/old.nix:75]: Currently `go mod vendor` prunes non-package directories https://github.com/golang/go/issues/26366, which caused various regressions  https://github.com/NixOS/nixpkgs/issues/89310 https://github.com/NixOS/nixpkgs/issues/89128 , and `vend` might not solve some cases https://github.com/NixOS/nixpkgs/pull/89453

So, I was thinking that we could use `go mod download` to generate the vendor directory, which seem to be the recommended way to vendorize https://github.com/go-modules-by-example/index/blob/master/012_modvendor/README.md. What do you think?

Comment (zowoq) [reply pkgs/development/go-modules/generic/old.nix:75]: Maybe I've misunderstood what you've proposing but I don't think the generated output of `download` is compatible with `-mod=vendor` so we'd need to work around it somehow which is basically what the original approach was doing?

Comment (marsam) [reply pkgs/development/go-modules/generic/old.nix:75]: > I don't think the generated output of download is compatible with -mod=vendor

oh, you are right! they aren't compatible. I thought `go mod download` could create a `vendor/`, sorry about the noise

**Review by kalbasit — APPROVED** [review]

**Comment by Mic92** [comment]
Yes. Actually I think should be not merge before the 20.09 branch off. So people that migrate from 20.03 to 20.09 have the time to migrate - They would get one release time to migrate everything which is the smallest time unit we can give them.

**Comment by zowoq** [comment]
> They would get one release time to migrate everything which is the smallest time unit we can give them.

Why?

**Comment by Mic92** [comment]
> > They would get one release time to migrate everything which is the smallest time unit we can give them.
> 
> Why?

This was not present in 20.03. If users upgrade to 20.09 all their packages needs an update. If you have a infrastructure large enough, you may not upgrade all installations at once.

**Comment by zowoq** [comment]
> This was not present in 20.03. If users upgrade to 20.09 all their packages needs an update. If you have a infrastructure large enough, you may not upgrade all installations at once.

Yes, it is a breaking change that is documented.

People will need to set `vendorSha256 = null` anyway to use `modSha256` so I don't see why we need to keep it around for another release cycle.

**Comment by Mic92** [comment]
Ok. Than we can merge it already.

**Comment by Mic92** [comment]
People that needs to support both can just have both checksums defined in the package.

File: pkgs/development/go-modules/generic/default.nix [hunk 2]
```diff
@@ -257,8 +255,5 @@ let
   });
 in if disabled then
   throw "${package.name} not supported for go ${go.meta.branch}"
-else if modSha256 != null then
-  lib.warn "modSha256 is deprecated and will be removed in the next release (20.09), use vendorSha256 instead" (
```
Comment (Mic92) [inline pkgs/development/go-modules/generic/default.nix:261]: Can you convert this in a throw statement that provides instruction to use `vendorSha256` instead?

Comment (zowoq) [reply pkgs/development/go-modules/generic/default.nix:261]: Changed to throw.

**Review by Mic92 — APPROVED** [review]

**Comment by kalbasit** [comment]
> People that needs to support both can just have both checksums defined in the package.

With the change to throw the above won't be true anymore. I think it'll be useful for people to support both. Can we throw only if modSha256 is defined but vendorSha256 is not?

**Comment by zowoq** [comment]
I'd rather we go back to the original which just outright removed `modSha256` without throws or warnings.

IMHO the error that `vendorSha256` is required is clear and it still allowed people to set `modSha256` if needed.

**Review by Mic92 — APPROVED** [review]

**Review by marsam — APPROVED** [review]
If you are going to merge it before the 20.09 branch-off, please update this bit: https://github.com/NixOS/nixpkgs/blob/c5815280e92112a25d958a2ec8b3704d7d90c506/nixos/doc/manual/release-notes/rl-2009.xml#L205-L207

**Comment by zowoq** [comment]
Oh, I didn't realise it said that.

It's been in the 20.09 release notes for a couple of months now, is it okay to change it?

**Pull request merged**